File: tier3_advanced/backtracking/problems/p039_combination_sum.py
"""P039: Combination Sum (Medium)

TODO: Implement solve() below.
Hint: Use backtracking with early pruning when the remaining sum goes negative.
"""
import sys
sys.path.insert(0, ".")
from src.utils import Problem, TestCase

class Solution(Problem):
    name = "39. Combination Sum"
    test_cases = [
        TestCase(input=([2,3,6,7], 7), expected=[[2,2,3],[7]], label="example 1"),
        TestCase(input=([2,3,5], 8), expected=[[2,2,2,2],[2,3,3],[3,5]], label="example 2"),
        TestCase(input=([2], 1), expected=[], label="no solution"),
    ]

    def solve(self, candidates: list[int], target: int) -> list[list[int]]:
        raise NotImplementedError("TODO: Implement solve(self, candidates: list[int], target: int) -> list[list[int]]")

if __name__ == "__main__":
    Solution().run()

File: tier3_advanced/backtracking/problems/p039_combination_sum.rs
/*
 * P39: Combination Sum (Medium)
 * https://leetcode.com/problems/combination-sum/
 * Topics: Array, Backtracking
 *
 * Given an array of distinct integers candidates and a target integer target, return a list of all unique combinations of candidates where the chosen numbers sum to target.
 * The same number may be chosen from candidates an unlimited number of times.
 *
 * Example 1:
 *     Input: candidates = [2,3,6,7], target = 7
 *     Output: [[2,2,3],[7]]
 *
 * Example 2:
 *     Input: candidates = [2,3,5], target = 8
 *     Output: [[2,2,2,2],[2,3,3],[3,5]]
 *
 * Hint: Use backtracking with early pruning when the remaining sum goes negative.
 */

use wasm_libs::*;

fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    todo!()
}

fn main() {
    let candidates = read_ints();
    let target = read_int();
    let result = combination_sum(candidates, target);
    for combo in result {
        write_ints(&combo);
    }
}
